### PR TITLE
Rework measurement part

### PR DESCRIPTION
**Pull request by ismael-mendoza**
This simplifies the measurement part of BTK taking into account suggestions in #100. Documentation still needs to be updated accordingly, but it would get to get some feedback before finalizing that. Thanks! 

@thuiop let me know if you have any additional suggestions. 

NOTE:  The diff is a little hard to understand so maybe looking at the file directly is better.

### CODE REVIEW

File: btk/measure.py
Comment (thuiop) [inline btk/measure.py:32]: I would rename the key in "deblended_images" for consistance. We'll also need to specify which format we want for all of the entries somewhere. Finally, we may want additionnal entries ? I'm thinking about segmentation in particular. Though we can probably handle in the metrics part the case where the key does not exist.

File: btk/measure.py
Comment (thuiop) [inline btk/measure.py:62]: Here we can easily get the segmentation with no additionnal effort, and use it to get some basic deblended images ; I believe it will be useful both as an example and as a way to test the metrics part later. I can do it if you prefer.

File: btk/measure.py
Comment (thuiop) [inline btk/measure.py:114]: I am unsure about which format to use exactly for the function. Shouldn't we only give it batch['blend_images'][index] ? Maybe adding the psf ? This does work so either way may be fine.

**Review by thuiop — COMMENTED** [review]
Great job ; I made some minor comments. We should probably get the opinion from someone else before merging (pro tip : go look directly at the file instead of the diff here, it will be much clearer)

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:114]: I think as long as we are clear what the batch contains it should be ok for the user. I'm just giving the user all the info assuming that there might be different algorithms that require different info

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:32]: Yeah do you have suggestions for specifying the format? I added some stuff in the top of the file but not sure if that is the best place. I could also add assertions checking the output of measurement in the `run_batch` method.

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:62]: I just added it, let me know if that makes sense (I had removed it and was planning on adding it back later anyways)

Comment (ismael-mendoza) [reply btk/measure.py]: This got me thinking, maybe we want to enforce that every algorithm returns a `peak` key because otherwise there is no way of telling how to match the `deblended_images` or `segmentation` to which isolated objects.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:37]: I think it should show up in the doc if it is at the top of the file, so it should be alright. Adding checks may help users troubleshooting their problems, this is a good idea.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py]: I think you can match even if you don't have the peaks, using intersection over union or something like that. It is easier to do with the peaks though.
The catalog idea is pretty good ; we should still make sure that people use standardized names but I think it would work great if we want to have support for photometric redshifts or shape measurement for instance.

Comment (thuiop) [reply btk/measure.py]: As for SEP, we can even have deblended_images by using the mask from the segmentation to recover the pixels corresponding to each object. I also think the raw segmentation from sep isn't a good format for segmentation in general as you can only have one object corresponding to each pixel ; we should instead have a (stamp_size,stamp_size,n_objects) boolean array.

Comment (ismael-mendoza) [reply btk/measure.py]: Thanks for reworking the SEP part, I agree with your reasoning. I'll try to add a catalog format with standard names.

**Comment by ismael-mendoza** [comment]
Hi @thuiop, I think this is ready for another review if you have sometime. Then we can discuss what we might be missing in terms of documentation/tutorials or if we want to do that separately. 

Once finalized, I'll ping Alexandre again to see if he can give us a final review.

File: btk/measure.py
Comment (thuiop) [inline btk/measure.py:17]: You should specify that dx and dy should be used when there is a single survey and ra and dec when you have multiresolution.

File: btk/measure.py
Comment (thuiop) [inline btk/measure.py:73]: We may want to specify that this function only works when you have a single survey.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:17]: Also, users should feel free to integrate any measurements they want in the catalog field, or even as other keys in the dictionnary ; those keys are mostly here to ensure we have a standardized format in the metrics, but they can have other measurements not covered by the metrics if they want.

**Review by thuiop — APPROVED** [review]
For me, this PR can be merged ; I have made a few small comments about the doc but overall it is ok for me.

**Review by thuiop — DISMISSED** [review]

**Comment by ismael-mendoza** [comment]
Waiting on a review from @aboucaud before merging. He said he can do it early next week, in the meantime I can start a branch with the metrics part based off of this branch.

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:538]: Why has the projection been removed from the docstring ?

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:60]: should be called coadd to be consistent with `sep_measure`.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:69]: Not very important but the naming `dx` and `dy` for centres is a bit puzzling.
Naturally I would expect something like `x_cen` or `x_peak`.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:91]: This needs a longer comment, since it can lead to errors down the line. 
One should probably add in the method docstring that the blend images are ndarray expected in a certain format.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:94]: Where does the 1.5 come from ?
Since it's hardcoded, it should at least be commented.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:96]:
```suggestion
    segmentation_exp = np.zeros((n_objects, stamp_size, stamp_size), dtype=bool)
```

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:97]:
```suggestion
    deblended_images = np.zeros((n_objects, *image.shape), dtype=image.dtype)
```

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:99]: is the `np.where()` needed here ?
Just a gut comment, haven't checked.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:94]: It's the threshold for detection against noise ; here it is set at 1.5 sigma. I agree that we should put up a comment.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:99]: Now that I think of it, probably not.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:141]: Not in this PR, but I think these two arguments are redundant.
```python
multiprocessing = True if cpu > 1 else False
```

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:175]:
```suggestion
            if not isinstance(out["catalog"], astropy.table.Table):
```

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:35]:
```suggestion
import astropy.table
```
It seems ones now needs to import the submodule `astropy.table` in order to access the Table, like `scipy`, at least on my computer. I'm running astropy==4.2.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:226]: I guess when you write such list comprehension, `blend_output` is copied multiple times, which is quite inefficient. 
Instead, a generator would be much better.
```suggestion
        input_args = ((blend_output, i) for i in range(self.batch_size))
```

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py:236]: What is the purpose of this ?

File: tests/test_measure.py
Comment (aboucaud) [inline tests/test_measure.py:49]: IMO, one should be able to extract two result values with much straightforward code.
This would mean refactoring a bit the output of the measurement methods.
Probably for a next PR.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:236]: We have a "list in, list out" approach for measure_functions ; if you provide a single function, you want to get your results directly, not as a list with a single element.

File: btk/survey.py
Comment (thuiop) [reply btk/survey.py:538]: Must be a mistake since next line is not removed.

File: btk/measure.py
Comment (thuiop) [reply btk/measure.py:91]: This should not be necessary since the images are supposed to come from a DrawBlendsGenerator directly, meaning they will be in the right format. When we make a tutorial for this, we should specify that images may come in the two different formats if users want to add their own measure_function.

File: btk/measure.py
Comment (aboucaud) [inline btk/measure.py]: I suggest something even more straightforward
```suggestion
        seg_i = segmentation == i + 1
        segmentation_exp[i] = seg_i
        deblended_images[i] = image * seg_i[np.newaxis, ...]
```

File: btk/measure.py
Comment (aboucaud) [reply btk/measure.py:237]: Then it's no more "list in, list out". And it makes subsequent tasks more difficult because you may need to handle cases where the measurement is not a list..

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:142]: Thanks @aboucaud , I created a new issue to address this.

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:237]: I think @aboucaud is right, handling this special case might be problematic. I wonder if that's an issue with the multiple survey code as well

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py]: Thanks I like this approach @aboucaud

File: btk/measure.py
Comment (ismael-mendoza) [reply btk/measure.py:69]: Yeah I think it can be puzzling too. I like your names better.

**Comment by ismael-mendoza** [comment]
Thanks everyone for your help ! I'm merging this now

**Pull request merged**
